### PR TITLE
8315810: Reimplement sun.reflect.ReflectionFactory::newConstructorForSerialization with method handles

### PR DESCRIPTION
**Pull request by mlchung**
This reimplements `sun.reflect.ReflectionFactory::newConstructorForSerialization` with method handles.

This API currently generates the bytecode which fails the verification because `new C; invokespecial A()` where the given class `C` and invoke a no-arg constructor of `C`'s first non-`Serializable` superclass `A` is not a valid operation per the VM specification. VM special cases the classes generated for reflection to skip verification for the constructors generated for serialization and externalization.  This change will allow such VM hack to be removed.

A `jdk.reflect.useOldSerializableConstructor` system property can be set to use the old implementation in case if customers run into any compatibility issue.   I expect this change has very low compatibility risk.   This system property is undocumented and will be removed in a future release.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8315811](https://bugs.openjdk.org/browse/JDK-8315811) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8315810](https://bugs.openjdk.org/browse/JDK-8315810): Reimplement sun.reflect.ReflectionFactory::newConstructorForSerialization with method handles (**Enhancement** - P3)
 * [JDK-8315811](https://bugs.openjdk.org/browse/JDK-8315811): Reimplement sun.reflect.ReflectionFactory::newConstructorForSerialization with method handles (**CSR**)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Contributors
 * Chen Liang `<liach@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15600/head:pull/15600` \
`$ git checkout pull/15600`

Update a local copy of the PR: \
`$ git checkout pull/15600` \
`$ git pull https://git.openjdk.org/jdk.git pull/15600/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15600`

View PR using the GUI difftool: \
`$ git pr show -t 15600`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15600.diff">https://git.openjdk.org/jdk/pull/15600.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15600#issuecomment-1708905352)

### CODE REVIEW

**Comment by ExE-Boss** [comment]
See also: [JDK‑8307575] ([GH‑13853])

[JDK‑8307575]: https://bugs.openjdk.org/browse/JDK-8307575 "[JDK‑8307575] Migrate the serialization constructor accessors to Method Handles"
[GH‑13853]: https://github.com/openjdk/jdk/pull/13853

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -3525,6 +3525,34 @@ public MethodHandle unreflectConstructor(Constructor<?> c) throws IllegalAccessE
             return lookup.getDirectConstructorNoSecurityManager(ctor.getDeclaringClass(), ctor);
         }
 
+        /*
+         * Produces a method handle that is capable of creating instances of the given class
+         * and instantiated by the given constructor.  No security manager check.
+         *
+         * This method is only used to implement the unsupported
+         * sun.reflect.ReflectionFactory::newConstructorForSerialization API.
+         */
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:3534]: The following is probably more correct:
```suggestion
        /**
         * Produces a method handle that is capable of creating instances of the given class
         * and instantiated by the given constructor.  No security manager check.
         *
         * This method is used to implement serialization and the unsupported
         * sun.reflect.ReflectionFactory::newConstructorForSerialization API.
         */
```

Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:3534]: What are you referring to "to implement serialization"?   It's only used by ReflectionFactory.

Comment (ExE-Boss) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:3534]: It’s called from the implementation of `JavaLangInvokeAccess::serializableConstructor`, which is called by `jdk.internal.reflect.ReflectionFactory::newConstructorForSerialization`, which is called by `java.io.ObjectStreamClass::getSerializableConstructor`:
https://github.com/openjdk/jdk/blob/86a18f5e2e0825dddb77656b2f43f64684f1464c/src/java.base/share/classes/java/io/ObjectStreamClass.java#L1444-L1446

Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:3534]: I see the confusion.   `newConstructionForSerialization` is used by serialization, exposed via `sun.reflect.ReflectionFactory`.   I may clarify it.

File: src/java.base/share/classes/jdk/internal/reflect/MethodHandleAccessorFactory.java [hunk 2]
```diff
@@ -101,23 +106,63 @@ static ConstructorAccessorImpl newConstructorAccessor(Constructor<?> ctor) {
...
+     * Creates a ConstructorAccessor that is capable of creating instances
+     * of the given class and instantiated by the given constructor.
+     *
+     * @param decl the class to instantiate
+     * @param ctor the constructor to call
+     * @return an accessible constructor
+     */
+    static ConstructorAccessorImpl newSerializableConstructorAccessor(Class<?> decl, Constructor<?> ctor) {
+        if (!constructorInSuperclass(decl, ctor)) {
+            throw new UnsupportedOperationException(ctor + " not a superclass of " + decl.getName());
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/reflect/MethodHandleAccessorFactory.java:127]: Notice in JShell 20, you could do this:
```
jshell> import sun.reflect.ReflectionFactory;
jshell> var rf = ReflectionFactory.getReflectionFactory()
jshell> var c = rf.newConstructorForSerialization(String.class, Boolean.class.getConstructor(boolean.class))
jshell> String d = (String) c.newInstance(false)
jshell> d.length()
```
which ends up in an NPE for `d.value` is null. This UOE is technically a new behavior that should be mentioned in the CSR.

Comment (mlchung) [reply src/java.base/share/classes/jdk/internal/reflect/MethodHandleAccessorFactory.java:127]: good catch.  will update the CSR.

File: src/java.base/share/classes/jdk/internal/reflect/DirectMethodHandleAccessor.java [hunk 3]
```diff
@@ -194,7 +194,8 @@ private boolean isIllegalArgument(RuntimeException ex) {
     private void checkReceiver(Object o) {
         // NOTE: will throw NullPointerException, as specified, if o is null
         if (!declaringClass.isAssignableFrom(o.getClass())) {
-            throw new IllegalArgumentException("object is not an instance of declaring class");
+            throw new IllegalArgumentException("object of type " + o.getClass().getName()
+                    +  " is not an instance of " + declaringClass.getName());
```
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/reflect/DirectMethodHandleAccessor.java:198]:
```suggestion
                    + " is not an instance of " + declaringClass.getName());
```

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 4]
```diff
@@ -3525,6 +3525,33 @@ public MethodHandle unreflectConstructor(Constructor<?> c) throws IllegalAccessE
...
+            if (decl == ctor.getDeclaringClass())
+                return true;
+
+            Class<?> cl = decl;
+            while ((cl = cl.getSuperclass()) != null) {
+                if (cl == ctor.getDeclaringClass()) {
+                    return true;
+                }
+            }
+            return false;
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:3552]: Does it break encapsulation too much to export and use the same method from jdk.internal.reflect.MethodHandleAccessorFactory.
Maybe not worth it for a small utility method.

Comment (mlchung) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:3552]: It's a small utility and I think it's ok to leave it this way.

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Pull request closed without merging**